source/adapters/cuda/command_buffer.cpp

```diff
@@ -261,11 +319,28 @@ static ur_result_t enqueueCommandBufferFillHelper(
       }
     }
 
+    CUgraphNode SignalNode = nullptr;
+    if (RetEvent) {
+      auto SignalEvent = CommandBuffer->addSignalNode(GraphNode, SignalNode);
+      *RetEvent = SignalEvent.release();
+    }
+
     // Get sync point and register the cuNode with it.
-    auto SyncPoint = CommandBuffer->addSyncPoint(GraphNode);
+    CUgraphNode SyncPointNode = SignalNode ? SignalNode : GraphNode;
+    auto SyncPoint = CommandBuffer->addSyncPoint(SyncPointNode);
     if (RetSyncPoint) {
       *RetSyncPoint = SyncPoint;
     }
+
+    std::vector<CUgraphNode> WaitNodes =
+        NumEventsInWaitList ? DepsList : std::vector<CUgraphNode>();
+    auto NewCommand = new T(CommandBuffer, GraphNode, SignalNode, WaitNodes);
```

Comment:
Would it be possible to use a `unique_ptr` instead of calling `new`?

Reply:
So this heap allocated object is returned to the user as an output parameter as handle typedefed to a raw pointer, which they can refcount with `release`/`retain` counts. So I don't think `unique_ptr` is the right abstraction because it is not singly owned by the runtime.

While ruminating about this, I think we don't actually need these objects to be user refcounted by the user, and can just have the same lifetime as the parent UR command-buffer (OpenCL does this already). At which point the command-buffer could store unique ptrs to these heap object. Making this change is out of scope for this PR, but I propose making it separately as a standalone PR? I can create an issue to track it

Reply:
That sounds like a good solution to me. 

FWIW in other UR entry points we tend to use `unique_ptr`s for heap allocations within entry points, just so resources are freed if something goes wrong at some stage within the UR entry point. At the end of the UR call we usually call `out_param = unique_ptr.release()` and give ownership of the mem to the user.